Loading SQL (Structured Query Language) extension for Jupyter

In [1]:
%load_ext sql

Connecting to the database

In [2]:
%sql mysql+pymysql://root:database_password@localhost:3306/md_water_services_new

Connecting to 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

## Data Inconsistency and Corruption 
This survey is aimed to confirm or denounce the susipicion made by the audit team. To check if the database has been tampered with and if employees are being truthful in their dealing with the people of Maji Ndogo. To also check if reports are being inputed correctly be the field surveyors.

In [3]:
%%sql

select *
from auditor_report

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

1620 rows affected.

+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
| location_id | type_of_water_source | true_water_source_score |                                                                 statements                                                                |
+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
|  SoRu34980  |         well         |            1            |          Residents admired the official's commitment to enhancing urban life, praising their cooperative and inclusive approach.          |
|  AkRu08112  |         well         |            3            | Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation. |
|  AkLu02044  |        river         |            0            |         Villagers were touched by the official's interactions, noting their humility, strong work ethic, and respectful attitude.         |
|  AkHa00421  |         well         |            3            |   Villagers were moved by the official's visit, praising their hard work, humility, and the profound sense of connection they fostered.   |
|  SoRu35221  |        river         |            0            |                 A photographer's lens captures the queue, though his own struggle for water is a hidden part of the story.                |
|  HaAm16170  |         well         |            1            |               With an open heart, the official created an atmosphere of unity and familial camaraderie among the villagers.               |
|  AkRu04812  |         well         |            3            |     The official's presence left an indelible mark, reflecting their humility, dedication, and the genuine connections they nurtured.     |
|  AkRu08304  |         well         |            3            |         The official's interactions resonated deeply with the villagers, leaving a lasting impression of respect and camaraderie.         |
|  AkRu05107  |         well         |            2            | Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation. |
|  AkRu05215  |         well         |            3            |                  Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern.                  |
+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
Truncated to displaylimit of 10.

From the previous task, i saw that the employees also took record of the water_source_score for each water sources in different locations in the water quality table. However, the column is named subjective_quality_score. To find out if the auditors scores and the employees score align, i decided to join both tables together.

In [7]:
%%sql 

SELECT * FROM water_quality;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

60146 rows affected.

+-----------+--------------------------+-------------+
| record_id | subjective_quality_score | visit_count |
+-----------+--------------------------+-------------+
|     0     |            0             |      1      |
|     1     |            1             |      1      |
|     2     |            5             |      1      |
|     3     |            10            |      1      |
|     4     |            4             |      1      |
|     5     |            0             |      1      |
|     6     |            9             |      1      |
|     7     |            10            |      1      |
|     8     |            2             |      1      |
|     9     |            10            |      1      |
+-----------+--------------------------+-------------+
Truncated to displaylimit of 10.

Quering both tables, i realized that they do not have a primary key that i can join them on. From the ERD created i saw that the visits table can be used to join both tables, which is the location_id and the record_id

In [8]:
%%sql

select *
from visits

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

60146 rows affected.

+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
| record_id | location_id |  source_id   |    time_of_record   | visit_count | time_in_queue | assigned_employee_id |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
|     0     |  SoIl32582  | SoIl32582224 | 2021-01-01 09:10:00 |      1      |       15      |          12          |
|     1     |  KiRu28935  | KiRu28935224 | 2021-01-01 09:17:00 |      1      |       0       |          46          |
|     2     |  HaRu19752  | HaRu19752224 | 2021-01-01 09:36:00 |      1      |       62      |          40          |
|     3     |  AkLu01628  | AkLu01628224 | 2021-01-01 09:53:00 |      1      |       0       |          1           |
|     4     |  AkRu03357  | AkRu03357224 | 2021-01-01 10:11:00 |      1      |       28      |          14          |
|     5     |  KiRu29315  | KiRu29315224 | 2021-01-01 10:17:00 |      1      |       9       |          40          |
|     6     |  AkRu05234  | AkRu05234224 | 2021-01-01 10:18:00 |      1      |       0       |          30          |
|     7     |  KiRu28520  | KiRu28520224 | 2021-01-01 10:28:00 |      1      |       0       |          34          |
|     8     |  HaZa21742  | HaZa21742224 | 2021-01-01 10:37:00 |      1      |       0       |          6           |
|     9     |  AmDa12214  | AmDa12214224 | 2021-01-01 10:58:00 |      1      |       0       |          36          |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
Truncated to displaylimit of 10.

In [17]:
%%sql

# Joining both tables on the visit table to get the location_id, subjective_quality_score, true_water_source_score and record_id

SELECT
    visits.record_id, auditor_report.location_id as audit_location, water_quality.subjective_quality_score,
    auditor_report.true_water_source_score, visits.location_id as visit_location
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

2698 rows affected.

+-----------+----------------+--------------------------+-------------------------+----------------+
| record_id | audit_location | subjective_quality_score | true_water_source_score | visit_location |
+-----------+----------------+--------------------------+-------------------------+----------------+
|    5185   |   SoRu34980    |            1             |            1            |   SoRu34980    |
|   59367   |   AkRu08112    |            3             |            3            |   AkRu08112    |
|   37379   |   AkLu02044    |            0             |            0            |   AkLu02044    |
|   51627   |   AkHa00421    |            3             |            3            |   AkHa00421    |
|   28758   |   SoRu35221    |            0             |            0            |   SoRu35221    |
|   31048   |   HaAm16170    |            1             |            1            |   HaAm16170    |
|    1513   |   AkRu04812    |            3             |            3            |   AkRu04812    |
|    1218   |   AkRu08304    |            3             |            3            |   AkRu08304    |
|    8322   |   AkRu05107    |            2             |            2            |   AkRu05107    |
|   21160   |   AkRu05215    |            10            |            3            |   AkRu05215    |
+-----------+----------------+--------------------------+-------------------------+----------------+
Truncated to displaylimit of 10.

In [18]:
%%sql

# To check if the location_id are the same in both tables

SELECT
    visits.record_id, auditor_report.location_id as audit_location, water_quality.subjective_quality_score,
    auditor_report.true_water_source_score, visits.location_id as visit_location
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
WHERE
    auditor_report.location_id != visits.location_id

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

+-----------+----------------+--------------------------+-------------------------+----------------+
| record_id | audit_location | subjective_quality_score | true_water_source_score | visit_location |
+-----------+----------------+--------------------------+-------------------------+----------------+
+-----------+----------------+--------------------------+-------------------------+----------------+

Since there are no details where the location in the audit table and the location in the visit table are not the same, i can move on and drop pon of the location_id columns. They both visited the exact locations to get records

In [19]:
%%sql

# To check if the location_id are the same in both tables

SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

2698 rows affected.

+-----------+-------------+----------------+-------------+
| record_id | location_id | employee_score | audit_score |
+-----------+-------------+----------------+-------------+
|    5185   |  SoRu34980  |       1        |      1      |
|   59367   |  AkRu08112  |       3        |      3      |
|   37379   |  AkLu02044  |       0        |      0      |
|   51627   |  AkHa00421  |       3        |      3      |
|   28758   |  SoRu35221  |       0        |      0      |
|   31048   |  HaAm16170  |       1        |      1      |
|    1513   |  AkRu04812  |       3        |      3      |
|    1218   |  AkRu08304  |       3        |      3      |
|    8322   |  AkRu05107  |       2        |      2      |
|   21160   |  AkRu05215  |       10       |      3      |
+-----------+-------------+----------------+-------------+
Truncated to displaylimit of 10.

Moving on, i checked if employee_score and the audit_score are the same

In [4]:
%%sql


SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
WHERE 
    water_quality.subjective_quality_score = auditor_report.true_water_source_score

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

2505 rows affected.

+-----------+-------------+----------------+-------------+
| record_id | location_id | employee_score | audit_score |
+-----------+-------------+----------------+-------------+
|    5185   |  SoRu34980  |       1        |      1      |
|   59367   |  AkRu08112  |       3        |      3      |
|   37379   |  AkLu02044  |       0        |      0      |
|   51627   |  AkHa00421  |       3        |      3      |
|   28758   |  SoRu35221  |       0        |      0      |
|   31048   |  HaAm16170  |       1        |      1      |
|    1513   |  AkRu04812  |       3        |      3      |
|    1218   |  AkRu08304  |       3        |      3      |
|    8322   |  AkRu05107  |       2        |      2      |
|   13070   |  HaDe16541  |       2        |      2      |
+-----------+-------------+----------------+-------------+
Truncated to displaylimit of 10.

Our results showed 2505 rows unlike when i ran the original query that showed 2689. Something is definitely wrong somewhere. Could it be that some people visited the a location more than once so the records are dupcated? I decided to bring in the visit count column and set it to show only when a location was visted once

In [6]:
%%sql

# To check if the water score are the same in both tables

SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, visits.visit_count
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
WHERE 
    water_quality.subjective_quality_score = auditor_report.true_water_source_score
AND visits.visit_count = 1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

1518 rows affected.

+-----------+-------------+----------------+-------------+
| record_id | location_id | employee_score | audit_score |
+-----------+-------------+----------------+-------------+
|    5185   |  SoRu34980  |       1        |      1      |
|   59367   |  AkRu08112  |       3        |      3      |
|   37379   |  AkLu02044  |       0        |      0      |
|   51627   |  AkHa00421  |       3        |      3      |
|   28758   |  SoRu35221  |       0        |      0      |
|   31048   |  HaAm16170  |       1        |      1      |
|    1513   |  AkRu04812  |       3        |      3      |
|    1218   |  AkRu08304  |       3        |      3      |
|    8322   |  AkRu05107  |       2        |      2      |
|   13070   |  HaDe16541  |       2        |      2      |
+-----------+-------------+----------------+-------------+
Truncated to displaylimit of 10.

We are making progress. 1518 results tally out of 1620 sites that the auditors visited.But that means that 102 records are incorrect.Using a WHERE operator clause, i checked if the scores in both tables are not the same and the locations where visisted just once.

In [7]:
%%sql

# To check if the water score are the same in both tables

SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

102 rows affected.

+-----------+-------------+----------------+-------------+
| record_id | location_id | employee_score | audit_score |
+-----------+-------------+----------------+-------------+
|   21160   |  AkRu05215  |       10       |      3      |
|    7938   |  KiRu29290  |       10       |      3      |
|   43140   |  KiHa22748  |       10       |      9      |
|   18495   |  SoRu37841  |       10       |      6      |
|   33931   |  KiRu27884  |       10       |      1      |
|   17950   |  KiZu31170  |       10       |      9      |
|   36864   |  KiZu31370  |       10       |      3      |
|   45924   |  AkRu06495  |       10       |      2      |
|   30524   |  HaRu17528  |       10       |      1      |
|   13192   |  SoRu38331  |       10       |      3      |
+-----------+-------------+----------------+-------------+
Truncated to displaylimit of 10.

Wowh! 102 scores do not align in both tables. Something is definitely wrong somewhere.

I decided to bring in the employee table to check the employees that made this errors since they assisgned scores incorrectly in some locations

I think there are two reasons this can happen.
1. These workers are all humans and make mistakes so this is expected.
2. Unfortunately, the alternative is that someone assigned scores incorrectly on purpose.

In either case, the employees are the source of the errors not the auditors, so i JOINED the assigned_employee_id for all the people on our list from the visits table to the query. This would show the employees name 

In [9]:
%%sql

SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, employee.employee_name 
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
JOIN
    employee 
ON
    employee.assigned_employee_id = visits.assigned_employee_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

102 rows affected.

+-----------+-------------+----------------+-------------+----------------+
| record_id | location_id | employee_score | audit_score | employee_name  |
+-----------+-------------+----------------+-------------+----------------+
|   21160   |  AkRu05215  |       10       |      3      |   Rudo Imani   |
|    7938   |  KiRu29290  |       10       |      3      |  Bello Azibo   |
|   43140   |  KiHa22748  |       10       |      9      |  Bello Azibo   |
|   18495   |  SoRu37841  |       10       |      6      |   Rudo Imani   |
|   33931   |  KiRu27884  |       10       |      1      |  Bello Azibo   |
|   17950   |  KiZu31170  |       10       |      9      | Zuriel Matembo |
|   36864   |  KiZu31370  |       10       |      3      | Yewande Ebele  |
|   45924   |  AkRu06495  |       10       |      2      |  Bello Azibo   |
|   30524   |  HaRu17528  |       10       |      1      | Jengo Tumaini  |
|   13192   |  SoRu38331  |       10       |      3      | Zuriel Matembo |
+-----------+-------------+----------------+-------------+----------------+
Truncated to displaylimit of 10.

#### Since the query is becoming long and massive, i decided to save it as a CTE so when i do more analysis, i can just call that CTE like it was a table.

I named the CTE incorrect_records

In [11]:
%%sql

WITH Incorrect_Records AS (SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, employee.employee_name 
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
JOIN
    employee 
ON
    employee.assigned_employee_id = visits.assigned_employee_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1)
SELECT *
FROM Incorrect_Records

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

102 rows affected.

+-----------+-------------+----------------+-------------+----------------+
| record_id | location_id | employee_score | audit_score | employee_name  |
+-----------+-------------+----------------+-------------+----------------+
|   21160   |  AkRu05215  |       10       |      3      |   Rudo Imani   |
|    7938   |  KiRu29290  |       10       |      3      |  Bello Azibo   |
|   43140   |  KiHa22748  |       10       |      9      |  Bello Azibo   |
|   18495   |  SoRu37841  |       10       |      6      |   Rudo Imani   |
|   33931   |  KiRu27884  |       10       |      1      |  Bello Azibo   |
|   17950   |  KiZu31170  |       10       |      9      | Zuriel Matembo |
|   36864   |  KiZu31370  |       10       |      3      | Yewande Ebele  |
|   45924   |  AkRu06495  |       10       |      2      |  Bello Azibo   |
|   30524   |  HaRu17528  |       10       |      1      | Jengo Tumaini  |
|   13192   |  SoRu38331  |       10       |      3      | Zuriel Matembo |
+-----------+-------------+----------------+-------------+----------------+
Truncated to displaylimit of 10.

To show the unique list of the employees from  the table and calculate how many mistakes each employee made. So basically i want to count how many times each employee name is in Incorrect_records list, and then group them by name using GROUP BY and using ORDER BY to check to the list in ascending order

In [12]:
%%sql

WITH Incorrect_Records AS (SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, employee.employee_name 
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
JOIN
    employee 
ON
    employee.assigned_employee_id = visits.assigned_employee_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1)
SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_Records
GROUP BY
    employee_name
ORDER BY
    number_of_mistakes DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

17 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
|   Rudo Imani   |         5          |
|   Farai Nia    |         4          |
|  Enitan Zuri   |         4          |
| Yewande Ebele  |         3          |
| Jengo Tumaini  |         3          |
|  Makena Thabo  |         3          |
+----------------+--------------------+
Truncated to displaylimit of 10.

Bello Azibo inputed 26 incorrect records? Malachi Mavusa 21? That's a lot to be referred to as a mistak. Something is fishy somewhere.

Let's say all employees make mistakes, if someone is corrupt, they will be making a lot of "mistakes", more than average, for example. But someone could just be clumsy, so i decided to not just conclude that they are all corrupt and tired to probe further

## Gathering Evidence

I saved the unique count query in a CTE and called it Error_count

In [21]:
%%sql

WITH Incorrect_Records AS (SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, employee.employee_name 
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
JOIN
    employee 
ON
    employee.assigned_employee_id = visits.assigned_employee_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1),

Error_count AS

(SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_Records
GROUP BY
    employee_name
ORDER BY
    number_of_mistakes DESC)
Select *
FROM Error_count

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

17 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
|   Rudo Imani   |         5          |
|   Farai Nia    |         4          |
|  Enitan Zuri   |         4          |
| Yewande Ebele  |         3          |
| Jengo Tumaini  |         3          |
|  Makena Thabo  |         3          |
+----------------+--------------------+
Truncated to displaylimit of 10.

I proceeded to calculate the average number of mistakes

In [24]:
%%sql

WITH Incorrect_Records AS (SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, employee.employee_name 
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
JOIN
    employee 
ON
    employee.assigned_employee_id = visits.assigned_employee_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1),

Error_count AS

(SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_Records
GROUP BY
    employee_name
ORDER BY
    number_of_mistakes DESC)
Select 
    AVG(number_of_mistakes) AS avg_error_count_per_empl
FROM Error_count

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

1 rows affected.

+--------------------------+
| avg_error_count_per_empl |
+--------------------------+
|          6.0000          |
+--------------------------+

I have to compare each employee's error_count with avg_error_count_per_empl and save the result as a CTE and call it Suspect_list.

#### To do this, i created a subquery in the WHERE clause

In [50]:
%%sql

WITH Incorrect_Records AS (SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, employee.employee_name 
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
JOIN
    employee 
ON
    employee.assigned_employee_id = visits.assigned_employee_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1),

Error_count AS

(SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_Records
GROUP BY
    employee_name
ORDER BY
    number_of_mistakes DESC),

Suspect_List AS
(SELECT
    employee_name, number_of_mistakes
FROM
    Error_count
WHERE
    number_of_mistakes > (SELECT AVG(number_of_mistakes)
                            FROM error_count))

Select *
FROM Suspect_List

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

4 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
+----------------+--------------------+

These 4 employees made mistakes that are above the average mistake i caluculated

To probe futher, i decided to bring in the statement column from the auditor table into the query since the auditors collected feedbacks from the people in each locations

#### The query is starting to look long and i want it to be easily understood and readable. So i coverted the Incorrect_record CTE to a VIEW. I can then use it as if it is a table. It would make my code much more simplier to read. After which i made Error_count a CTE

In [25]:
%%sql
CREATE VIEW
    Incorrect_Records AS (SELECT
    visits.record_id, auditor_report.location_id as location_id, water_quality.subjective_quality_score as employee_score,
    auditor_report.true_water_source_score as audit_score, employee.employee_name,auditor_report.statements AS statements
FROM
    auditor_report
JOIN
    visits
ON
    auditor_report.location_id = visits.location_id
JOIN
    water_quality
ON
    water_quality.record_id = visits.record_id
JOIN
    employee 
ON
    employee.assigned_employee_id = visits.assigned_employee_id
WHERE 
    water_quality.subjective_quality_score != auditor_report.true_water_source_score
AND visits.visit_count = 1)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

++
||
++
++

In [73]:
%%sql

SELECT *
FROM Incorrect_Records
LIMIT 5

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

5 rows affected.

+-----------+-------------+----------------+-------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------+
| record_id | location_id | employee_score | audit_score | employee_name |                                                               statements                                                               |
+-----------+-------------+----------------+-------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------+
|   21160   |  AkRu05215  |       10       |      3      |   Rudo Imani  |                Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern.                 |
|    7938   |  KiRu29290  |       10       |      3      |  Bello Azibo  |            A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.             |
|   43140   |  KiHa22748  |       10       |      9      |  Bello Azibo  |                  A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system.                   |
|   18495   |  SoRu37841  |       10       |      6      |   Rudo Imani  | The official's respectful and diligent presence was met with heartfelt appreciation, creating a sense of closeness with the villagers. |
|   33931   |  KiRu27884  |       10       |      1      |  Bello Azibo  |                 A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community.                 |
+-----------+-------------+----------------+-------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------+

In [72]:
%%sql

WITH Error_count AS

(SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_Records
GROUP BY
    employee_name
ORDER BY
    number_of_mistakes DESC),

Suspect_List AS
    (SELECT employee_name, number_of_mistakes
FROM error_count
WHERE
    number_of_mistakes > (SELECT AVG(number_of_mistakes)
                            FROM error_count))
Select *
FROM Suspect_List

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

4 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
+----------------+--------------------+

Recap ...
1. I used Incorrect_records to find all of the records where the auditor and employee scores don't match.
2. I then used error_count to aggregate the data, and got the number of mistakes each employee made.
3. Finally, suspect_list retrieves the data of employees who make an above-average number of mistakes.

Now I can filter the Incorrect_records VIEW to identify all of the records associated with the four employees we identified in the suspect list.

In [68]:
%%sql

WITH error_count
    AS #This CTE calculates the number of mistakes each employee made
(SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
     Incorrect_records
#Incorrect_records is a view that joins the audit report to the database for records where the auditor and employees scores are different
GROUP BY
    employee_name),
Suspect_List AS
    (SELECT employee_name, number_of_mistakes
FROM error_count
WHERE
    number_of_mistakes > (SELECT AVG(number_of_mistakes)
                            FROM error_count))
SELECT employee_name, location_id, statements
FROM Incorrect_Records
WHERE employee_name IN (SELECT employee_name FROM Suspect_List)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

71 rows affected.

+----------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name  | location_id |                                                                                         statements                                                                                        |
+----------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  Bello Azibo   |  KiRu29290  |                                      A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.                                      |
|  Bello Azibo   |  KiHa22748  |                                            A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system.                                            |
|  Bello Azibo   |  KiRu27884  |                                           A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community.                                          |
| Zuriel Matembo |  KiZu31170  |                                     A community leader stood with his people, expressing concern for the water quality and the time lost in queues.","                                    |
|  Bello Azibo   |  AkRu06495  |                                        A healthcare worker in the queue expressed fears about water-borne diseases, her face etched with worry.","                                        |
| Zuriel Matembo |  SoRu38331  | An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts. |
| Malachi Mavuso |  AmAm09607  |       Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.      |
| Zuriel Matembo |  AkHa00314  |                                    A street vendor's sales suffer from time spent waiting, her concern for the water's quality affecting her products.                                    |
| Malachi Mavuso |  KiRu26598  |                                      A teenager's dreams are tempered by reality, her future threatened by the corrupt practices she sees around her.                                     |
|  Bello Azibo   |  KiIs23853  |       Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception.      |
+----------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
Truncated to displaylimit of 10.

If you have a look, you will notice some alarming statements about these four officials (look at these records: KiIs23853,  for example. See how the word "cash" is used in the statements. I decided to filter the records that refer to "cash"

In [70]:
%%sql

WITH error_count
    AS #This CTE calculates the number of mistakes each employee made
(SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
     Incorrect_records
#Incorrect_records is a view that joins the audit report to the database for records where the auditor and employees scores are different
GROUP BY
    employee_name),
Suspect_List AS
    (SELECT employee_name, number_of_mistakes
FROM error_count
WHERE
    number_of_mistakes > (SELECT AVG(number_of_mistakes)
                            FROM error_count))
SELECT employee_name, location_id, statements
FROM Incorrect_Records
WHERE employee_name IN (SELECT employee_name FROM Suspect_List)
AND statements LIKE "%cash%"


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

19 rows affected.

+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name  | location_id |                                                                                              statements                                                                                              |
+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Zuriel Matembo |  SoRu38331  |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
| Malachi Mavuso |  AmAm09607  |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
|  Bello Azibo   |  KiIs23853  |            Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception.            |
|  Bello Azibo   |  HaSe21323  |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
| Zuriel Matembo |  AkRu05880  |               Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The allusion to cash changing hands deepened their skepticism.               |
|  Bello Azibo   |  KiRu27065  |           Villagers expressed their discomfort with an official who displayed a haughty demeanor and negligence. The mention of cash transactions deepened their growing sense of unease.            |
| Malachi Mavuso |  KiRu25347  |              Villagers expressed their discontent with an official who appeared dismissive and neglectful. The mention of cash changing hands added to their growing sense of distrust.              |
| Zuriel Matembo |  SoIl32575  | Villagers recounted unsettling encounters with an official known for their arrogance and avoidance of responsibilities. The mention of cash changing hands added to their apprehension and distrust. |
|  Bello Azibo   |  AkRu04508  |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
| Lalitha Kaburi |  AkRu07310  |                Villagers spoke of their unsettling encounters with an official who seemed indifferent and uninterested, hinting at potential improprieties involving cash exchanges.                 |
+----------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
Truncated to displaylimit of 10.

#### WOWH!! Suspicion confirmed! These 4 employees actually requested from cash in some location. Bribery and corruption at the highest level

Just to be sure that there are not other employees that requested for cash, I filtered the other result that are not in the suspect table and they mentioned cash

In [71]:
%%sql

WITH error_count
    AS #This CTE calculates the number of mistakes each employee made
(SELECT
    employee_name, COUNT(employee_name) AS number_of_mistakes
FROM
     Incorrect_records
#Incorrect_records is a view that joins the audit report to the database for records where the auditor and employees scores are different
GROUP BY
    employee_name),
Suspect_List AS
    (SELECT employee_name, number_of_mistakes
FROM error_count
WHERE
    number_of_mistakes > (SELECT AVG(number_of_mistakes)
                            FROM error_count))
SELECT employee_name, location_id, statements
FROM Incorrect_Records
WHERE employee_name NOT IN (SELECT employee_name FROM Suspect_List)
AND statements LIKE "%cash%"

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

+---------------+-------------+------------+
| employee_name | location_id | statements |
+---------------+-------------+------------+
+---------------+-------------+------------+

I get an empty result, so no one, except the four suspects, has these allegations of bribery.

So i can sum up the evidence i have for Zuriel Matembo, Malachi Mavuso, Bello Azibo and Lalitha Kaburi:
1. They all made more mistakes than their peers on average.
2. They all have incriminating statements made against them, and only them.